### PR TITLE
changefeedccl: support TIMESTAMP and DECIMAL with the avro format

### PR DESCRIPTION
**Pull request by danhhz**
Timestamp is relatively straightforward and uses the corresponding
logical type.

The avro spec also has a logical type for decimals, but ours cover a
number of cases that it doesn't (NaN, Inf, per-datum precision and
scale). I think the right answer for them is to make this a union with a
`crdb.decimal` logical type that has a string representation.
Unfortunately, we can't do that without some surgery to the avro library
we use and it seems too late in the game to do that. For now, leave them
unsupported.

A number of column types are still unsupported with avro, but this is
probably a good enough core that we can mention the experimental
support.

Release note (enterprise change): CHANGEFEEDs now have limited and
experimental support for the avro format.

### CODE REVIEW

**Comment by danhhz** [comment]
@mjibson can you look at decimalToRat/ratToDecimal and the TestDecimalRatRoundtrip test? I think I managed to get the decimal to roundtrip through the crazy avro decimal bytes encoding for the special case of a fixed scale. Left the other cases as a TODO.

**Review by mrtracy — COMMENTED** [review]
This is soupy, but I think that's just going to come with the territory for now.  I think that testing this through a schema change might be prudent; it seems that a column could be added which would cause errors in the serialization step. I believe we already handle those errors, but it still seems like a more complicated situation than we're currently covering.

Otherwise, its LGTM.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31104#-:-LOKdjNSD_x2Y1BEGYs7:b-2osliv)* status: :shipit: complete! 0 of 0 LGTMs obtained

File: pkg/ccl/changefeedccl/avro.go [hunk 1]
```diff
@@ -281,8 +351,61 @@ func (r *avroSchemaRecord) rowFromNative(native interface{}) (sqlbase.EncDatumRo
...
+// precision is set) this is roundtripable without information loss.
+//
+// TODO(dan): We really should just be controlling our own encoding destiny
+// here. Make that possible.
+func decimalToRat(dec apd.Decimal, scale int32) (big.Rat, error) {
+	if dec.Form != apd.Finite {
+		return big.Rat{}, errors.Errorf(`cannot convert %s form decimal: %s`, dec.Form, &dec)
+	}
+	if e := -dec.Exponent; e > scale {
+		return big.Rat{}, errors.Errorf(`%s has more than %d digits after the decimal`, &dec, scale)
```
Comment (mjibson) [inline pkg/ccl/changefeedccl/avro.go]: Is this check required in addition to the next one? That is, is there a test case that would trigger this first error but not the second one? Or is it just to have a clearer error message?

File: pkg/ccl/changefeedccl/avro_test.go [hunk 2]
```diff
@@ -343,3 +381,50 @@ func TestAvroMigration(t *testing.T) {
...
+			{1, apd.New(0, -1)},
+			{1, apd.New(1, -1)},
+			{1, apd.New(123, -1)},
+			{5, apd.New(1, -5)},
+		}
+		for d, test := range tests {
+			rat, err := decimalToRat(*test.dec, test.scale)
+			require.NoError(t, err)
+			roundtrip := ratToDecimal(rat, test.scale)
+			if test.dec.Cmp(&roundtrip) != 0 {
```
Comment (mjibson) [inline pkg/ccl/changefeedccl/avro_test.go]: You may want to use `CmpTotal` here which does exact comparison and will catch differences with trailing zeros.

**Review by madelynnblue — APPROVED** [review]
Decimal stuff LGTM

**Review by danhhz — COMMENTED** [review]
TFTRs!

>   I think that testing this through a schema change might be prudent; it seems that a column could be added which would cause errors in the serialization step.

Done

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31104#-:-LOOfV9v8PPzpU6whq1z:byp9i9d)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ccl/changefeedccl/avro.go, line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31104#-LOKMSHlvFCj4zs7GOfJ-r1-374:-LOO_e_4D0D3iKe4ks9u:b-bi9rd) ([raw file](https://github.com/cockroachdb/cockroach/blob/33ef01cfa4b2ab1226379f8442eccdbdfdc5de20/pkg/ccl/changefeedccl/avro.go#L374)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Is this check required in addition to the next one? That is, is there a test case that would trigger this first error but not the second one? Or is it just to have a clearer error message?
</blockquote></details>

Mostly an artifact of my exploratory programming here. Removed
___
*[pkg/ccl/changefeedccl/avro_test.go, line 406 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31104#-LOKMSHlvFCj4zs7GOfK-r1-406:-LOOYGt_7_g-DbWssRWy:bloxpgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/33ef01cfa4b2ab1226379f8442eccdbdfdc5de20/pkg/ccl/changefeedccl/avro_test.go#L406)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

You may want to use `CmpTotal` here which does exact comparison and will catch differences with trailing zeros.
</blockquote></details>

Done. Looks like I used it below but missed this one. Had to remove two test cases that didn't meet the contract, which seems fine

**Comment by danhhz** [comment]
bors r=mjibson,mrtracy

**Pull request merged**